### PR TITLE
Add implementation of focus metric

### PR DESCRIPTION
**Pull request by joannasendorek**
## Description

Implementation of focus metric #70 

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes.

- [x] added unit tests
- [x] new example notebook

## Checklist:

- [x] I have updated all related `pyproject.toml` files.
- [x] I have updated lock files.
- [x] I have set an Assignee and Reviewers.
- [x] I have labeled the PR correctly.
- [x] I have performed a self-review of my own code.
- [x] I have covered my code and changes with unit tests.
- [x] I have annotated types as extensively as possible.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [x] All code references and dependencies will work.

### CODE REVIEW

File: example/notebooks/focus_metric_example.ipynb [hunk 1]
```diff
@@ -0,0 +1,379 @@
...
+   "cell_type": "markdown",
+   "metadata": {
+    "tags": []
+   },
+   "source": [
+    "## Demo for general algorithms "
+   ]
+  },
+  {
+   "cell_type": "code",
```
Comment (adamwawrzynski) [inline example/notebooks/focus_metric_example.ipynb:253]: You can delete this cell - it's not used.

File: example/notebooks/focus_metric_example.ipynb [hunk 2]
```diff
@@ -0,0 +1,379 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Download `YOLOv5` and `ImageNet.yaml` files from https://github.com/ultralytics/yolov5 if not present in local storage."
+   ]
+  },
+  {
+   "cell_type": "code",
```
Comment (adamwawrzynski) [inline example/notebooks/focus_metric_example.ipynb:106]: You can delete this cell - it's not used.

File: example/notebooks/focus_metric_example.ipynb [hunk 3]
```diff
@@ -0,0 +1,379 @@
...
+   "cell_type": "markdown",
+   "metadata": {
+    "tags": []
+   },
+   "source": [
+    "#### Downloading missing models"
+   ]
+  },
+  {
+   "cell_type": "markdown",
```
Comment (adamwawrzynski) [inline example/notebooks/focus_metric_example.ipynb:99]: You can delete this cell - it's not obsolete.

File: example/notebooks/focus_metric_example.ipynb [hunk 4]
```diff
@@ -0,0 +1,379 @@
...
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "os.environ['CUDA_LAUNCH_BLOCKING'] = \"1\""
+   ]
+  },
+  {
+   "cell_type": "markdown",
```
Comment (adamwawrzynski) [inline example/notebooks/focus_metric_example.ipynb:90]: You can delete this cell - it's obsolete.

File: example/notebooks/focus_metric_example.ipynb [hunk 5]
```diff
@@ -0,0 +1,379 @@
...
+   "metadata": {},
+   "source": [
+    "Install missing libraries required by `YOLOv5` that are not part of `foxai` package."
+   ]
+  },
+  {
+   "cell_type": "code",
```
Comment (adamwawrzynski) [inline example/notebooks/focus_metric_example.ipynb:65]: You can delete this cell - those dependencies are not required.

File: example/notebooks/focus_metric_example.ipynb [hunk 6]
```diff
@@ -0,0 +1,379 @@
...
+    "import torch\n",
+    "from torch.utils.data import DataLoader\n",
+    "import torchvision\n",
+    "\n",
+    "from foxai.context_manager import FoXaiExplainer, ExplainerWithParams, CVClassificationExplainers\n",
+    "from foxai.visualizer import mean_channels_visualization"
+   ]
+  },
+  {
+   "cell_type": "markdown",
```
Comment (adamwawrzynski) [inline example/notebooks/focus_metric_example.ipynb:58]: You can delete this cell - it's obsolete.

File: example/notebooks/focus_metric_example.ipynb [hunk 7]
```diff
@@ -0,0 +1,379 @@
...
+   "metadata": {},
+   "source": [
+    "# Example notebook for Image Classification (with imagenet-mini)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "In this tutorial, You will see how You can use explainable algorithms to study pre-trained model decision. We will take a pre-trained model, sample images and run several explainable methods."
```
Comment (adamwawrzynski) [inline example/notebooks/focus_metric_example.ipynb:14]: Update this description.

File: example/notebooks/focus_metric_example.ipynb [hunk 8]
```diff
@@ -0,0 +1,379 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Example notebook for Image Classification (with imagenet-mini)"
```
Comment (adamwawrzynski) [inline example/notebooks/focus_metric_example.ipynb:7]: Update this description.

File: example/notebooks/focus_metric_example.ipynb [hunk 9]
```diff
@@ -0,0 +1,379 @@
...
+   },
+   "source": [
+    "#### Define custom functions"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Define function that will load model, list of labels and transformation function of a desired model. Currently we support, in this notebook, only a few models: `VGG11`, `ResNet50`, `ViT`, `MobileNetV3` and `YOLOv5`. You can easilly add new models from `torchvision` model zoo and even define Your own model."
```
Comment (adamwawrzynski) [inline example/notebooks/focus_metric_example.ipynb:129]: Update description. In the notebook, you are using only one network architecture.

**Review by adamwawrzynski — COMMENTED** [review]
Overall looks good but I've tested this metric with 2 explainers: `LayerGradCAM` and `LayerDeepLIFT` and there seems to be a problem with the normalization/standardization of attributes before calculating the `Focus` metric. Maybe you should add this step before calculating the score.

GradCAM Focus metric: ~0.91
GradCAM attributes mean value: ~0.15
GradCAM figure:
gradcam_focus

DeepLIFT Focus metric: ~0.39
DeepLIFT mean attributes: ~5e-05
DeepLIFT figure:
layer_deeplift_focus

File: foxai/metrics/focus.py [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+        mosaics_num: A number of mosaics to generate.
+        target_class: Target class to be used for mosaics creation.
+        pandas_random_state: a random state to be passed to DataFrame.sample() method, use for repetitive results
+
+    Returns:
+        A list of MosaicData objects which contain 2x2 images mosaic and a 2x2 matrices of their labels.
+    """
+    mosaic_data_list = []
+    for _ in range(mosaics_num):
+        df = pd.DataFrame(images_with_labels, columns=["image", "label"])
```
[the remark this answers is not in the document]
Comment (joannasendorek) [reply foxai/metrics/focus.py:72]: I get your simplicity argument and I can re-write it to dictionary although then we have to keep track of: each label mapped to a list of pictures, then randomly choosing labels and randomly choosing from list of pictures. Pandas code is far more efficient for large datasets, if we want to sample from thousands of images. So the question is if we should prioritize efficiency or simplicity here?

File: foxai/metrics/focus.py [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+        mosaic_image: Torch Tensor representing image where each quadrant is coming from different source image.
+        mosaic_labels: Torch Tensor of 2x2 dimension representing labels of each quadrant image.
+    """
+
+    mosaic_image: torch.Tensor
+    mosaic_labels: torch.Tensor
+
+
+def create_mosaic_picture(images_with_labels: List[List[torch.Tensor]]) -> MosaicData:
+    """Create mosaic from a list of four pictures. Images are placed at random positions.
```
[the remark this answers is not in the document]
Comment (joannasendorek) [reply foxai/metrics/focus.py:24]: This is a good idea, although I wonder if in this case we shouldn't given user full control, meaning setting `shuffle` as a flag and if set to False, use pictures order from the list? This way not only target class pictures could be controlled, but also non-target.

File: example/notebooks/focus_metric_example.ipynb (hunk 4 above)
Comment (joannasendorek) [reply example/notebooks/focus_metric_example.ipynb:76]: It relates to custom function for model loading, which is used later? Have I got it properly?

File: foxai/visualizer.py [hunk 12]
```diff
@@ -21,12 +19,49 @@
...
+def draw_image(
+    image: torch.Tensor,
+    title: str = "",
+    ax: Optional[Axes] = None,
+    figsize: Tuple[int, int] = (8, 8),
+) -> Figure:
+    """Draw picture.
+
+    Args:
+        image: Image in shape (C x H x W).
```
[the remark this answers is not in the document]
Comment (joannasendorek) [reply foxai/visualizer.py:31]: I think they could also be integers, since in line 44 I do .as_type(float)

File: foxai/metrics/focus.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (adamwawrzynski) [reply foxai/metrics/focus.py:74]: Are you sure Pandas is faster? I've heard a lot about its inefficiency in handling large datasets. Could you compare those two aproaches?

File: foxai/metrics/focus.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (adamwawrzynski) [reply foxai/metrics/focus.py:26]: Yes, that's a good idea.

File: foxai/visualizer.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (adamwawrzynski) [reply foxai/visualizer.py:31]: Ok, fair point.

File: example/notebooks/focus_metric_example.ipynb (hunk 4 above)
Comment (adamwawrzynski) [reply example/notebooks/focus_metric_example.ipynb]: You already removed this cell.

File: foxai/metrics/focus.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (joannasendorek) [reply foxai/metrics/focus.py:28]: Implemented in https://github.com/softwaremill/FoXAI/pull/94/commits/316dfe28111ff70e316c0b3fcfb6d5cd98f23a59

File: foxai/metrics/focus.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (joannasendorek) [reply foxai/metrics/focus.py]: Implemented in: https://github.com/softwaremill/FoXAI/pull/94/commits/03c49d75d8ad512b133075acf1a7d6443c282f59

**Review by adamwawrzynski — APPROVED** [review]
LGTM
